### PR TITLE
Add functional tests

### PR DESCRIPTION
**Pull request by imstar15**
**Unit test**

```
npm run test
```

**Functional test**

We need to to setup a SENDER_MNEMONIC environment variable for the test and ensure that there are enough tokens in this account (recommended the balance is greater than or equals to 30 UNIT).

```
SENDER_MNEMONIC="<SENDER_MNEMONIC>" npm run test:functional
```

**Test all**

```
SENDER_MNEMONIC="<SENDER_MNEMONIC>" npm run test:all
```

### CODE REVIEW

File: src/scheduler.ts [hunk 1]
```diff
@@ -168,7 +168,7 @@ export class Scheduler {
    * @param sendingAddress 
    * @param receivingAddress 
    */
-  validateTransferParams(amount: number, sendingAddress: string, receivingAddress: string): void {
+  validateTransferParams(amount: number, sendingAddress: AddressOrPair, receivingAddress: string): void {
```
Comment (justinzhou93) [inline src/scheduler.ts:171]: ah thanks for the catch!

File: src/constants.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ export const ADDITIONAL_UNIT = 1
 export const NO_DIFF = 0
 export const RECURRING_TASK_LIMIT = 24
 export const LOWEST_TRANSFERRABLE_AMOUNT = 1000000000
+export const SS58_PREFIX = 51;
```
Comment (justinzhou93) [inline src/constants.ts:10]: oh good add!

File: src/scheduler.ts [hunk 3]
```diff
@@ -190,13 +190,16 @@ export class Scheduler {
...
     const txObject = polkadotApi.tx(extrinsicHex)
     const unsub = await txObject.send(async (result) => {
+      const { status } = result;
       if (_.isNil(handleDispatch)) {
         await this.defaultErrorHandler(result)
       } else {
         await handleDispatch(result)
       }
+      if (status.isFinalized) {
+        unsub();
```
Comment (justinzhou93) [inline src/scheduler.ts:200]: hmmm can it call unsub() inside itself?

File: test/functional.test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import _ from 'lodash';
```
Comment (justinzhou93) [inline test/functional.test.ts:1]: are these the e2es?

File: test/functional.test.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  expect(section).toEqual('automationTime');
+  expect(method).toEqual('scheduleNativeTransferTask');
+  expect(providedIdOnChain).toEqual(providedID);
+  expect(receiverAddressOnChain.toString()).toEqual(receiverAddress);
+  expect(amountOnChain.toNumber()).toEqual(amount);
+  const isTimestampsEqual = _.reduce(executionTimestamps, (prev, executionTimestamp, index) => prev && executionTimestamp === executionTimestampsOnChain[index].toNumber(), true);
+  expect(isTimestampsEqual).toEqual(true);
+})
+
+test('scheduler.buildScheduleNotifyExtrinsic works', async () => {});
```
Comment (justinzhou93) [inline test/functional.test.ts:64]: don't forget to add these 2 tests!

**Review by justinzhou93 — COMMENTED** [review]
couple of small things, but overall looks good for first cut! thanks for running through this!

File: src/scheduler.ts (hunk 3 above)
Comment (imstar15) [reply src/scheduler.ts:200]: Yes.
See polkadot.api examples.
https://polkadot.js.org/docs/api/start/api.tx.subs

File: test/functional.test.ts (hunk 4 above)
Comment (imstar15) [reply test/functional.test.ts:1]: Does e2es means End-to-End Solution?

This test uses oak-js-api, neumann chain, polkadot.js api.

I think it should be considered e2e.

Comment (irsal) [reply test/functional.test.ts:1]: e2e's means end-to-end test. So it touches multiple integrations and parts of the stack

Comment (irsal) [reply test/functional.test.ts:1]: https://www.digitalocean.com/community/tutorials/how-to-write-end-to-end-tests-in-node-js-using-puppeteer-and-jest

Comment (imstar15) [reply test/functional.test.ts:1]: Okay, I will study it well.

**Review by justinzhou93 — COMMENTED** [review]
hmmm all of these tests locally are failing for me. is there some special way for me to run these to make them work?

**Review by justinzhou93 — APPROVED** [review]
can you update `SENDER_MNEMONIC="<SENDER_MNEMONIC>" node_modules/.bin/jest --runInBand` as a new command in package json and then split out npm test to only run non-e2e tests? i.e. right now `npm test` fails, but i would like for that to succeed. but other than that, lgtm

**Review by irsal — APPROVED** [review]
good to merge

**Pull request merged**